extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java

```diff
@@ -1253,186 +1278,218 @@ static String getParameterName(MethodInfo method, int position) {
+                    if (it.hasNext()) {
+                        // Delegate to the next "resolve_x" method
+                        bc.return_(bc.invokeVirtual(ClassMethodDesc.of(ClassDesc.of(generatedClassName),
+                                resolveMethodPrefix + "_resolve_" + (idx + 1),
+                                CompletableFuture.class, String.class,
+                                EvaluatedParams.class, CompletableFuture.class),
+                                bundleCreator.this_(), name, evaluatedParams, bundleCreator.this_()));
+                    } else {
+                        // Last group - return null
+                        bc.returnNull();
+                    }
+                });
+            });
         }
-    }
 
-    private void addMessageMethod(MethodCreator resolve, String key, MessageMethod method, ResultHandle name,
-            ResultHandle evaluatedParams,
-            ResultHandle ret, String bundleClass) {
-        List<Type> methodParams = method.parameterTypes();
-
-        BytecodeCreator matched = resolve.ifTrue(Gizmo.equals(resolve, resolve.load(key), name))
-                .trueBranch();
-        if (methodParams.isEmpty()) {
-            matched.invokeVirtualMethod(Descriptors.COMPLETABLE_FUTURE_COMPLETE, ret,
-                    method.isMessageBundleInterfaceMethod()
-                            ? matched.invokeInterfaceMethod(method.descriptor(), matched.getThis())
-                            : matched.invokeVirtualMethod(method.descriptor(), matched.getThis()));
-            matched.returnValue(ret);
-        } else {
-            // The CompletionStage upon which we invoke whenComplete()
-            ResultHandle paramsReady = matched.readInstanceField(Descriptors.EVALUATED_PARAMS_STAGE,
-                    evaluatedParams);
-
-            FunctionCreator whenCompleteFun = matched.createFunction(BiConsumer.class);
-            matched.invokeInterfaceMethod(Descriptors.CF_WHEN_COMPLETE, paramsReady, whenCompleteFun.getInstance());
-
-            BytecodeCreator whenComplete = whenCompleteFun.getBytecode();
+        bundleCreator.method("resolve", mc -> {
+            mc.returning(CompletionStage.class);
+            ParamVar evalContext = mc.parameter("ec", EvalContext.class);
+
+            mc.body(bc -> {
+                LocalVar name = bc.localVar("name", bc.invokeInterface(Descriptors.GET_NAME, evalContext));
+                LocalVar ret = bc.localVar("ret", bc.new_(CompletableFuture.class));
+
+                // First handle dynamic messages, i.e. the "message" virtual method
+                bc.if_(new ObjectOps(bc, name).equals_(Const.of(MESSAGE)), dynamicMessage -> {
+                    Expr evaluatedMessageKey = dynamicMessage.invokeStatic(Descriptors.EVALUATED_PARAMS_EVALUATE_MESSAGE_KEY,
+                            evalContext);
+                    Expr paramsReady = dynamicMessage.get(evaluatedMessageKey.field(Descriptors.EVALUATED_PARAMS_STAGE));
```

Comment:
You don't have to call `get()` at all, just use `evaluatedMessageKey.field(Descriptors.EVALUATED_PARAMS_STAGE)` directly.